### PR TITLE
[stable/kong] For Cassandra: Fix install migration and add basic wait-for-db

### PR DESCRIPTION
**Pull request by stackedsax**
Signed-off-by: stackedsax <ascammon@hotmail.com>

#### What this PR does / why we need it:

I tried the helm install with cassandra as the db and it failed to run the migrations cleanly.  The process would 'work' only when I: 

1.  run the `helm install`, wait until the `migrations` template is 'completed' and the kong deployment is stuck in the 'Init' phase
2. run a `helm upgrade` to kick off the `migrations-pre-upgrade` and `migrations-post-upgrade` templates

I'd be happy to be told that I'm doing something wrong, but if this is what others are seeing it's clearly not ideal.

After some digging, the problem was that even though Cassandra's 9042 port is open, it's not quite ready to run the `migration bootstrap`.  The bootstrap would start but in the middle of the migration the `migrations` template errors and restarts.  When it spins up the second time, it thinks that the bootstrap is finished and goes to a 'Completed' state.  

Indeed, running `kong migrations bootstap` at this point returns 'database already bootstrapped'.  However, running `kong migrations list` returns:
```
executed migrations:
core: 000_base, 001_14_to_15, 002_15_to_1
oauth2: 000_base_oauth2, 001_14_to_15, 002_15_to_10
acl: 000_base_acl
database has pending migrations:
acl: 001_14_to_15
database has new migrations available:
jwt: 000_base_jwt, 001_14_to_15
basic-auth: 000_base_basic_auth, 001_14_to_15
key-auth: 000_base_key_auth, 001_14_to_15
rate-limiting: 000_base_rate_limiting, 001_14_to_15, 002_15_to_10
hmac-auth: 000_base_hmac_auth, 001_14_to_15
response-ratelimiting: 000_base_response_rate_limiting, 001_14_to_15, 002_15_to_10
run 'kong migrations up' to proceed
```
So, it looked to me like the failed bootstrap needed to be 'finished' and then another migration run and finished before the database would be ready to go.

Even though it may look as ugly as sin, I added a `&& kong migrations finish && kong migrations up && kong migrations finish` to the command that the `migrations` template runs.  Now, when the `migrations` template spins up after the first error, it finishes the bootstrap and runs the rest of the migrations.  Additionally, this doesn't affect the postgres installation negatively.

In the process of debugging this, I tried various approaches: 

- I added an initContainer to all of the migrations templates hoping that simply waiting for port 9042 to become active would help.  It didn't.  
- I tried raising the various `db-timeout` and `lock-timeout` parameters for `kong migrations` but no dice.  
- I tried waiting for a cqlsh command to return successfully, but that yielded the same behavior as the netcat version and also required a significant chunk of ugly code to be run as a single command to install cqlsh in the first place.  
- Lastly, I considered installing nodetool but that would require cassandra's JMX port to be open to external connections.  This, in turn, requires a number of environment variables to be set and to modify the base cassandra image by copying in a jmx password file, a la:
   - https://gist.github.com/franzwong/a7f36d7e45076f85f3832dc0c2dc0c0e (This seemed out of scope... and I also wasn't completely convinced that `nodetool status` wouldn't report the cluster up too early as well)

So after all of that, I fell back to the somewhat ugly but predictable method outlined above.  I left in the netcat initContainer stuff because I thought that was a reasonable thing to do.

#### Special notes for your reviewer:
If it helps in your testing, I used something like this to switch back and forth between databases :

`helm install -n testkong01 . --set env.database=cassandra --set cassandra.enabled=true --set postgresql.enabled=false`

##### Variables Documentation
 
I think there are a bunch of other things to clean up in this chart around the cassandra parameters.  I'd like to put that work in a separate PR, if that's alright.

For example, I used the parameter `cassandra.config.ports.cql` as the KONG_CASSANDRA_CQL_PORT since that is the parameter that the Cassandra chart uses.  I notice the kong chart suggests a `cassandra_port` parameter, but I don't see it used in any template.  Likewise, the `env.cassandra_repl_factor` is not used anywhere.  

Meanwhile, the `cassandra_contact_points` parameter is not used in any template but my guess is that it should be used whenever the environment variable `KONG_CASSANDRA_CONTACT_POINTS` is set so that the contact points aren't merely `{{ template "kong.cassandra.fullname" . }}`.  For example:

- https://github.com/helm/charts/blob/master/stable/kong/templates/deployment.yaml#L60

I thought removing or adding or changing parameter names might require more discussion so I wanted a different PR to deal with that.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [ ] Variables are documented in the README.md

### CODE REVIEW

**Comment by stackedsax** [comment]
@hbagdi This is a continuation of your PR: 

- https://github.com/helm/charts/pull/9047

Let me know if I've missed something really obvious, but I hope this is helpful.

**Comment by maorfr** [comment]
Hey @stackedsax,

Can you please prepend the PR title with `[stable/kong]` ?

Thanks!

**Comment by stackedsax** [comment]
@maorfr done. Thanks -- didn't notice that was the convention around here.

**Comment by maorfr** [comment]
Not always, but it will help draw attention from kong maintainers, instead of cassandra maintainers ;)

File: stable/kong/templates/migrations.yaml
Comment (hbagdi) [inline stable/kong/templates/migrations.yaml:64]: What is going on here?
`kong migrations bootstrap` should be sufficient for initial migration.

File: stable/kong/templates/migrations-post-upgrade.yaml
Comment (hbagdi) [inline stable/kong/templates/migrations-post-upgrade.yaml]: At the end of the day, we are checking if a specific port is listening for new TCP connections or not.
It makes sense to not duplicate these initContainer images but rather change the value of the env vars.

Meaning, we define `DB_HOST` and `DB_PORT` and then use that in the `nc` command.
The value of those env vars can be a smaller `if else` block. Thoughts?

**Review by hbagdi — CHANGES_REQUESTED** [review]

File: stable/kong/templates/migrations.yaml
Comment (stackedsax) [reply stable/kong/templates/migrations.yaml:64]: > kong migrations bootstrap should be sufficient for initial migration.

That's just it: it isn't.  

As I mentioned:

>The bootstrap would start but in the middle of the migration the migrations template errors and restarts. When it spins up the second time, it thinks that the bootstrap is finished and goes to a 'Completed' state.

The end result was a 'Completed' kong-init-migrations pod but a half-finished bootstrap.

Looking back at my logs, this is what I would see in the `kong-init-migrations` pod when I do an initial run of a `helm install -n testkong01 kong`:

```
$ kubectl logs -f testkong01-kong-init-migrations-xv7jj
2019/01/12 23:06:32 [warn] 8#0: *2 [lua] cluster.lua:496: refresh(): [lua-cassandra] found contact point with '0.0.0.0' as rpc_address, using 'testkong01-cassandra' to contact it instead. If this is incorrect you should avoid using '0.0.0.0' in rpc_address, context: ngx.timer
bootstrapping database...
Error: [Cassandra error] failed to insert cluster lock: [Write timeout] Operation timed out - received only 0 responses.

  Run with --v (verbose) or --vv (debug) for more details
rpc error: code = Unknown desc = Error: No such container: 1cfa7813c602fa226d0e66c935f9fb396667e0cccef8592d455d4709dc9df37d
```
The container exits due to the error.  

The container will restart and this time says:
```
$ kubectl logs -f testkong01-kong-init-migrations-xv7jj 
2019/01/12 23:09:32 [warn] 8#0: *2 [lua] cluster.lua:496: refresh(): [lua-cassandra] found contact point with '0.0.0.0' as rpc_address, using 'testkong01-cassandra' to contact it instead. If this is incorrect you should avoid using '0.0.0.0' in rpc_address, context: ngx.timer
database already bootstrapped
```

File: stable/kong/templates/migrations-post-upgrade.yaml
Comment (stackedsax) [reply stable/kong/templates/migrations-post-upgrade.yaml]: Hm, we could reduce things down to one initContainer, but since this chart can enable both the postgres and cassandra databases at the same time it makes things complicated.  To construct the correct command for each of the three options will probably require a helper template to be defined.  Can you think of an easier option to handle creating a bash command for the three options: postgres/cassandra/both?

To illustrate my point, this is a reasonably ugly way of doing what I'm describing:
```
{{- if or (.Values.postgresql.enabled) (.Values.cassandra.enabled)  }}
command: [ "/bin/sh", "-c", "until [[ nc -zv $KONG_PG_HOST $KONG_PG_PORT -w1 && $KONG_CASSANDRA_HOST $KONG_CASSANDRA_PORT -w1]]; do echo 'waiting for dbs'; sleep 1; done" ]
{{- end }}
{{- else .Values.postgresql.enabled }}
command: [ "/bin/sh", "-c", "until nc -zv $KONG_PG_HOST $KONG_PG_PORT -w1; do echo 'waiting for db'; sleep 1; done" ]
{{- end }}
{{- else .Values.cassandra.enabled }}
command: [ "/bin/sh", "-c", "until nc -zv $KONG_CASSANDRA_HOST $KONG_CASSANDRA_PORT -w1; do echo 'waiting for db'; sleep 1; done" ]
{{- end }}
```
I don't know the exact syntax for the first option, but it's something like that.

I think a helper template could output one of those three bash commands and then use that function in this template.  Let me know if that's the direction you think this should go, or if there's an easier alternative I'm not seeing.

Comment (stackedsax) [reply stable/kong/templates/migrations-post-upgrade.yaml]: On further thought, perhaps the `nc` could only address the database that is set with the `env.database`.  Both databases may be enabled, but only the one that is being used will need to be bootstrapped.  This might simplify things a bit.  

It could all be done without a helper template, but because the port and host parameters are not simply `DB_HOST` and `DB_PORT`, there's still a bunch of if/else's to set a local variable based on which database is being used.  It's probably still clearer if it's taken care of in a helper template.

Comment (hbagdi) [reply stable/kong/templates/migrations-post-upgrade.yaml]: While functionally it is possible to enable Postgres and Cassandra, enabling both the databases at the same time doesn't make any sense, so we can safely assume that they will never be enabled at the same time. We can use documentation to warn anyone to not do such a thing.

Comment (stackedsax) [reply stable/kong/templates/migrations-post-upgrade.yaml]: It's true in practice, but perhaps this phrasing in the README should be a little more severe:

> Enabling both will create both databases in your cluster, but only one will be used by Kong based on the env.database parameter. Postgres is enabled by default.

Would it be worth spending a bit of time and enforcing that only one db is ever at a time?  It might solve some confusion for people using this chart.

As for the work to remove the duplicate initContainer, I still think it would be worth hiding the complexity of building the right command in a helper template.  If the variables were simply `DB_HOST` and `DB_PORT` for both databases it would be relatively simple to insert them into the command.  Instead, to do it within the `migrations` template I think we'd need an if/else block that sets the `DB_HOST` and `DB_PORT` as local variables from either the postgres or cassandra database values and then use those in the command.  It's fine in either place, it just might be prettier to hide that chunk of code in a helper template.  Do you have a preference?

Comment (hbagdi) [reply stable/kong/templates/migrations-post-upgrade.yaml]: Putting it in a helper template is in my todo list but haven't got around it, so if you do it, that would be great. Let's hae a small if block around postgres, cassandra for the value of DB_HOST and DB_PORT.

Also, I'll be off the grid for few weeks so please expect delayed responses. Thank you!

Comment (stackedsax) [reply stable/kong/templates/migrations-post-upgrade.yaml]: @hbagdi Hope you had some pleasant off-the-grid time.

I simplified the initContainer block with some helper templates.  Let me know what you think.

I also removed the KONG_PG_PASSWORD variable entirely from the initContainer since netcat doesn't use it.

I could even remove the environment variables entirely from the initContainer section and simply define the entire command in a helper template.  However, I thought the way I checked in was simple enough and might actually be nice for people to see what was going on in the template.  I could go either way, just let me know.

File: stable/kong/templates/migrations.yaml
Comment (stackedsax) [reply stable/kong/templates/migrations.yaml:64]: @hbagdi Let me know your thoughts on this one.  

Maybe it's just a matter of my Cassandra having fewer resources than normal in development and therefore not being immediately available even if netcat says the port is open.  However, if I'm having this problem with a several-node cluster, I bet others will run into the same problem if they're trying to take kong for a spin in minikube and want to use Cassandra there.

Comment (hbagdi) [reply stable/kong/templates/migrations.yaml:64]: @stackedsax apologies for the delay on our side. Things have been busy and I was out for a couple weeks.

An update: This is a bug in Kong where it detects migrations to be complete when they are not, migrations need to be reentrant.  We're tracking the bug internally currently and will be working to fix it.

To unblock this, we can proceed with this but let's add a `TODO` comment to change this back to `kong migrations bootstrap` once Kong includes the fix.

Comment (hbagdi) [reply stable/kong/templates/migrations.yaml:64]: Friendly ping on this. Please update the PR when you can, we would like this to get this in, thanks!

Comment (stackedsax) [reply stable/kong/templates/migrations.yaml:64]: Will do -- been trying to find the time to get back to this.  If I don't get to it tonight, then by the end of the week.

**Comment by stackedsax** [comment]
@hbagdi Sorry for not getting back to this ever.  I think you just wanted a comment telling people that this would be changed soon?  Has it already been fixed or would you like me to create another PR with more of a comment explaining what this is doing?

**Comment by hbagdi** [comment]
Please open another PR, thanks.

**Pull request closed without merging**
